
---
title: "Haplomics report for `r params$ASSAY` at *`r params$LOCUS`*."
params:
  date: !r Sys.Date()
  LOCUS: ""
  ASSAY: ""
  hist: ""
  anot: ""
  hap1: ""
  hap2: ""
  heat: ""
  res:  ""
  summ: ""

execute:
  cache: false
  freeze: auto  # optional, disables reuse of old knit.md if in a workflow
output: 
  html_notebook:
    theme: journal #united #flatly
    #highlight: #espresso #tango 
    toc: TRUE
    toc_depth: 5
    toc_float: 
      collapsed: TRUE
      smooth_scroll: TRUE
      number_sections: FALSE
format: 
  revealjs:
    transition: slide,
    background-transition: fade
editor: visual
geometry: margin=0.95in
---

This report is generated by Quarto on `r format(Sys.Date(), "%B %d, %Y")`


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, include=FALSE}

# Loading libraries
library(knitr)
library(rmarkdown)
library(tidyverse)
library(DT)

#library(cowplot)
#library(ggplot2)
```

|
|
|

### Introduction
This report demonstrates results of the conducted analysis for `r params$ASSAY` at *`r params$LOCUS`*.

|
|
|

### Minor allele frequency of variants

```{r MAF, echo=FALSE, fig.align = "center", out.width='70%', fig.cap="Figure 1. Allele frequency of the variants", fig.subcap=c('a','b', 'c')}

knitr::include_graphics(params$hist)
```

|
|
|


### Variants annotation plot

```{r annotation, echo=FALSE, fig.align="center", out.width='80%', fig.cap="Figure 2. Number and annotation of the variants in each gene"}

knitr::include_graphics(params$anot)
```

|
|
|

#### Discriptive statistics 


```{r summary of the datasets, echo=FALSE, message=FALSE, warning=TRUE}

#data_summ <- read.delim(file = params$summ, header = TRUE, sep = "\t", comment.char = "")

#data_summ %>% knitr::kable("pipe", caption = "Table 1. Discriptive statistics of merged datasets")

# Read report from file
report_contents <- readLines(params$summ)

# Print report to console
#cat(report_contents, sep = "\n")

```
`r report_contents`

|
|
|

### Haplotype plot 

#### Haplotype variants

```{r haplotypes, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='100%', fig.cap="Figure 3. Alleles distribution across haplotypes"}

knitr::include_graphics(params$hap1)
```

|
|
|

#### Haplotype tagging variants

```{r haplotypes narrowed, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='100%', fig.cap="Figure 4. Haplotypes narrowed by varied variants"}

knitr::include_graphics(params$hap2)
```

|
|
|

### Associations with `r params$ASSAY`

#### Heatmap plot

```{r heatmap traits, echo=FALSE, fig.align="center", out.width='100%', fig.cap="Figure 5. Heatmap plot of the associations coefficients"}

knitr::include_graphics(params$heat)
```

|
|
|

#### Table of results

```{r significant results traits, echo=FALSE, message=FALSE, warning=TRUE}

result <- readRDS(params$res)

# show the results
result %>% DT::datatable(caption = "Table 1. Results of haplotype association")
```

|
|
|
