
---
title: "Haplomics report for `r params$ASSAY` at *`r params$LOCUS`*."
params:
  date: !r Sys.Date()
  LOCUS: ""
  ASSAY: ""
  hist: ""
  anot: ""
  hap1: ""
  hap2: ""
  heat: ""
  res:  ""
  summ: ""

execute:
  cache: false
output: 
  html_notebook:
    theme: journal #united #flatly
    #highlight: #espresso #tango 
    toc: TRUE
    toc_depth: 5
    toc_float: 
      collapsed: TRUE
      smooth_scroll: TRUE
      number_sections: FALSE
format: 
  revealjs:
    transition: slide,
    background-transition: fade
editor: visual
geometry: margin=0.95in
---

This report is generated by Quarto on `r format(Sys.Date(), "%B %d, %Y")`


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, include=FALSE}

# Loading libraries
library(knitr)
library(rmarkdown)
library(tidyverse)
library(DT)

#library(cowplot)
#library(ggplot2)
```

|
|
|

### Introduction
This report demonstrates results of the conducted analysis for `r params$ASSAY` at *`r params$LOCUS`*.

|
|
|

### Minor allele frequency of variants

```{r MAF, echo=FALSE, fig.align = "center", out.width='70%', fig.cap="Figure 2. Allele frequency of the variants", fig.subcap=c('a','b', 'c')}

print(params$hist)
file.exists(params$hist)
knitr::include_graphics(params$hist)
```

|
|
|


### Variants annotation plot

```{r annotation, echo=FALSE, fig.align="center", out.width='80%', fig.cap="Figure 4. Number and annotation of the variants in each gene"}

knitr::include_graphics(params$anot)
```

|
|
|

#### Discriptive statistics 


```{r summary of the datasets, echo=FALSE, message=FALSE, warning=TRUE}

#data_summ <- read.delim(file = params$summ, header = TRUE, sep = "\t", comment.char = "")

#data_summ %>% knitr::kable("pipe", caption = "Table 1. Discriptive statistics of merged datasets")

# Read report from file
report_contents <- readLines(params$summ)

# Print report to console
#cat(report_contents, sep = "\n")

```
`r report_contents`

|
|
|

### Results for `r params$ASSAY` 

#### Haplotype plot

```{r haplotypes, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='100%', fig.cap="Figure 5. Alleles distribution across haplotypes"}

knitr::include_graphics(params$hap1)
```

|
|
|

#### Haplotype tagging variants

```{r haplotypes narrowed, figures-side, echo=FALSE, fig.show="hold", fig.align="center", out.width='100%', fig.cap="Figure 6. Haplotypes narrowed by varied variants"}

knitr::include_graphics(params$hap2)
```

|
|
|

#### Heatmap plot of haplotype associations with `r params$ASSAY`

```{r heatmap traits, echo=FALSE, fig.align="center", out.width='100%', fig.cap="Figure 7. Heatmap plot of haplotype associations"}

knitr::include_graphics(params$heat)
```

|
|
|

#### Table of significant associations

```{r significant results traits, echo=FALSE, message=FALSE, warning=TRUE}

#res_phen    <- str_split(params$res, " ", simplify = TRUE)[,1]
#result_phen <- read.delim(file = res_phen, header = TRUE, sep = ",", comment.char = "")
result <- readRDS(params$res)

# show the results
result %>% DT::datatable(caption = "Table 2. Results of haplotype association") # with `r params$ASSAY`
```

|
|
|


```{r significant results protein, echo=FALSE, message=FALSE, warning=TRUE}

#res_prot    <- str_split(params$res, " ", simplify = TRUE)[,3]
##result_prot <- read.delim(file = res_prot, header = TRUE, sep = ",", comment.char = "")
#result_prot <- readRDS(res_prot)

#result_sig <- result_sig %>% arrange(p.value, haplo)

# change columns names
##names(result_sig) <- names(result_sig) %>% str_remove("X\\.+\\d\\.")

# show the results
#result_prot %>% DT::datatable(caption = "Table 4. Results of haplotypes association with plasma proteins")

```



```{r res, echo=FALSE}
# Display the table

#DT::datatable(result_sig, caption = "Results Table", options = list(pageLength = 10))
#%>% knitr::kable_styling(position = "center") 
#find /scratch/mfilosi -type f -name "*.Rmd"

#knitr::kable(
#  result_sig,
  #"simple",
  #"latex",
  #digits = 5,
  #format.args = list(scientific = TRUE), 
#  full_width = FALSE,
#  html_font = "Cambria", 
#  caption = "Results Table of Association with Clinical Traits"
#)

```